File: scripts/close_old_prs.py
import requests
import os
from datetime import datetime, timedelta
import argparse

# GitHub repository details
REPO_OWNER = "vectordotdev"
REPO_NAME = "vector"

# GitHub personal access token (read from environment variable)
GITHUB_TOKEN = os.getenv("GITHUB_TOKEN")
print(GITHUB_TOKEN[-10:])
if not GITHUB_TOKEN:
    raise EnvironmentError("GITHUB_TOKEN environment variable is not set.")

# Headers for authentication
HEADERS = {
    "Authorization": f"Bearer {GITHUB_TOKEN}",
    "Accept": "application/vnd.github+json",
}

# Define the cutoff date (6 months ago)
CUTOFF_DATE = datetime.now() - timedelta(days=6 * 30)  # Approximation for 6 months

# GitHub API endpoint for listing pull requests
PULLS_URL = f"https://api.github.com/repos/{REPO_OWNER}/{REPO_NAME}/pulls"

COMMENT = (
    "Thank you for your contribution to Vector! To keep the repository tidy and focused, we are closing this PR due to inactivity. "
    "We greatly appreciate the time and effort you've put into this PR."
    "If you'd like to continue working on it, we encourage you to re-open the PR and we would be delighted to review it again. "
    "Before re-opening, please use `git merge origin master` to resolve any conflicts with origin/master."
)

# Fetch pull requests
def fetch_pull_requests():
    params = {
        "state": "open",
        "per_page": 100
    }
    response = requests.get(PULLS_URL, headers=HEADERS, params=params)
    response.raise_for_status()
    return response.json()

# Add a comment to a pull request
def add_comment_to_pr(pr_number, comment):
    url = f"https://api.github.com/repos/{REPO_OWNER}/{REPO_NAME}/issues/{pr_number}/comments"
    data = {"body": comment}
    response = requests.post(url, headers=HEADERS, json=data)
    if response.status_code != 200:
        print(f"Request Headers: {response.headers}")
        print(f"Request Body: {response.json()}")
        print(f"Failed to close PR #{pr_number}. Response status: {response.status_code}")
    response.raise_for_status()
    print(f"Added comment to PR #{pr_number}")

# Close a pull request
def close_pull_request(pr_number):
    url = f"https://api.github.com/repos/{REPO_OWNER}/{REPO_NAME}/pulls/{pr_number}"
    data = {"state": "closed"}
    response = requests.patch(url, headers=HEADERS, json=data)
    response.raise_for_status()
    print(f"Closed PR #{pr_number}")

# Main function
def main(dry_run):
    print("Fetching pull requests...")
    pull_requests = fetch_pull_requests()

    closed_prs = []

    for pr in pull_requests:
        pr_number = pr["number"]
        pr_title = pr["title"]
        created_at = datetime.strptime(pr["created_at"], "%Y-%m-%dT%H:%M:%SZ")

        if created_at < CUTOFF_DATE:
            labels = [label["name"] for label in pr["labels"]]

            if "meta: awaiting author" in labels:
                print(f"PR #{pr_number} (created at: {created_at}) to be closed.")
                if not dry_run:
                    add_comment_to_pr(pr_number, COMMENT)
                    close_pull_request(pr_number)
                closed_prs.append({"number": pr_number, "title": pr_title, "created_at": created_at})

    print("\nReport:")
    print(f"Total PRs that would be closed: {len(closed_prs)}" if dry_run else f"Total PRs closed: {len(closed_prs)}")
    for pr in closed_prs:
        print(f"- PR #{pr['number']}: {pr['title']} (Created at: {pr['created_at']})")

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Close inactive GitHub pull requests.")
    parser.add_argument("--dry-run", action="store_true", help="If set, do not modify any PRs, only print what would be done.")
    args = parser.parse_args()

    main(dry_run=args.dry_run)

File: src/commands/close_old_prs.rs
use crate::config::Config;
use anyhow::{Context, Result};
use chrono::{DateTime, Duration, Utc};
use reqwest::blocking::Client;
use serde_json::Value;

const AWAITING_LABEL: &str = "meta: awaiting author";
const CLOSE_COMMENT: &str = "\
Thank you for your contribution to Vector! To keep the repository tidy and focused, \
we are closing this PR due to inactivity. \
We greatly appreciate the time and effort you've put into this PR. \
If you'd like to continue working on it, we encourage you to re-open the PR and we would be delighted to review it again. \
Before re-opening, please git merge origin master to resolve any conflicts with origin/master.";

pub fn run(config: &Config, dry_run: bool, yes: bool) -> Result<()> {
    let client = Client::new();
    let cutoff = Utc::now() - Duration::days(6 * 30);

    println!("Fetching open PRs older than {}...", cutoff.date_naive());
    let prs = fetch_open_prs(&client, config)?;
    println!("Total open PRs fetched: {}", prs.len());

    // Collect matching PRs before acting
    let mut matched: Vec<(u64, String, DateTime<Utc>)> = Vec::new();

    for pr in &prs {
        let number = pr["number"].as_u64().unwrap_or(0);
        let title = pr["title"].as_str().unwrap_or("");
        let created_at = pr["created_at"].as_str().unwrap_or("");
        let created_dt = created_at.parse::<DateTime<Utc>>().unwrap_or(Utc::now());

        if created_dt >= cutoff {
            continue;
        }

        let labels: Vec<&str> = pr["labels"]
            .as_array()
            .map(|a| a.iter().filter_map(|l| l["name"].as_str()).collect())
            .unwrap_or_default();

        if labels.contains(&AWAITING_LABEL) {
            matched.push((number, title.to_string(), created_dt));
        }
    }

    if matched.is_empty() {
        println!("No matching PRs found.");
        return Ok(());
    }

    println!("\nFound {} PR(s) to close:", matched.len());
    for (number, title, created_at) in &matched {
        println!(
            "  PR #{number}: {title} (created: {})",
            created_at.date_naive()
        );
    }

    if dry_run {
        println!("\n[dry-run] No changes made.");
        return Ok(());
    }

    if !crate::confirm(&format!("Close {} PR(s)?", matched.len()), yes) {
        println!("Aborted.");
        return Ok(());
    }

    for (number, _, _) in &matched {
        add_comment(&client, config, *number, CLOSE_COMMENT)?;
        close_pr(&client, config, *number)?;
    }

    println!("\nClosed {} PR(s).", matched.len());
    Ok(())
}

fn fetch_open_prs(client: &Client, config: &Config) -> Result<Vec<Value>> {
    let url = format!(
        "https://api.github.com/repos/{}/{}/pulls",
        config.repo_owner, config.repo_name
    );
    let mut prs = Vec::new();
    let mut page = 1u32;

    loop {
        let resp = client
            .get(&url)
            .header("Authorization", format!("Bearer {}", config.github_token))
            .header("Accept", "application/vnd.github.v3+json")
            .header("User-Agent", "github-tools")
            .query(&[
                ("state", "open"),
                ("per_page", "100"),
                ("page", &page.to_string()),
            ])
            .send()
            .context("Failed to fetch PRs")?
            .error_for_status()?;

        let batch: Vec<Value> = resp.json()?;
        let done = batch.len() < 100;
        prs.extend(batch);
        if done {
            break;
        }
        page += 1;
    }
    Ok(prs)
}

fn add_comment(client: &Client, config: &Config, pr_number: u64, body: &str) -> Result<()> {
    let url = format!(
        "https://api.github.com/repos/{}/{}/issues/{pr_number}/comments",
        config.repo_owner, config.repo_name
    );
    client
        .post(&url)
        .header("Authorization", format!("Bearer {}", config.github_token))
        .header("Accept", "application/vnd.github.v3+json")
        .header("User-Agent", "github-tools")
        .json(&serde_json::json!({"body": body}))
        .send()?
        .error_for_status()
        .with_context(|| format!("Failed to add comment to PR #{pr_number}"))?;
    println!("Added comment to PR #{pr_number}");
    Ok(())
}

fn close_pr(client: &Client, config: &Config, pr_number: u64) -> Result<()> {
    let url = format!(
        "https://api.github.com/repos/{}/{}/pulls/{pr_number}",
        config.repo_owner, config.repo_name
    );
    client
        .patch(&url)
        .header("Authorization", format!("Bearer {}", config.github_token))
        .header("Accept", "application/vnd.github.v3+json")
        .header("User-Agent", "github-tools")
        .json(&serde_json::json!({"state": "closed"}))
        .send()?
        .error_for_status()
        .with_context(|| format!("Failed to close PR #{pr_number}"))?;
    println!("Closed PR #{pr_number}");
    Ok(())
}
